varnish-macros/src/parser_args.rs

```diff
@@ -180,34 +179,44 @@ impl ParamType {
         } else {
             // Only standard types left, possibly optional
             not_in! { Event, "Event functions can only have `Ctx`, `#[event] Event`, and `#[shared_per_vcl] &mut Option<Box<T>>` arguments." }
-            if let Some(arg_type) = as_option_type(arg_ty).and_then(ParamTy::try_parse) {
-                // Special "always-optional" types should be stored as non-optional
-                let dflt = Self::get_default(pat_ty, arg_type)?;
-                Self::Value(ParamInfo::new(arg_type, dflt, true))
-            } else if let Some(arg_type) = ParamTy::try_parse(arg_ty) {
-                if arg_type.must_be_optional() {
-                    error! { "This type of argument must be declared as optional with `Option<...>`" }
+            let (opt, arg_ty) =
+                if let Some(arg_ty) = as_option_type(arg_ty).and_then(ParamTy::try_parse) {
+                    (true, arg_ty)
+                } else if let Some(arg_ty) = ParamTy::try_parse(arg_ty) {
+                    (false, arg_ty)
+                } else {
+                    Err(error(&pat_ty, "unsupported argument type"))?
+                };
+            if !opt && arg_ty.must_be_optional() {
+                error! { "This type of argument must be declared as optional with `Option<...>`" }
+            }
+            let default = Self::get_arg_opts(pat_ty, arg_ty)?;
+            let has_required = Self::get_required_attr(pat_ty)?;
+            let opt = if has_required {
+                if !opt {
+                    error! { "The `required` attribute is only allowed on Option<...> arguments" }
                 }
-                let dflt = Self::get_default(pat_ty, arg_type)?;
-                Self::Value(ParamInfo::new(arg_type, dflt, false))
+                if !arg_ty.must_be_optional() {
+                    error! { "The `required` attribute is only allowed on Probe, ProbeCow, and SocketAddr arguments" }
```

Comment:
that one is confusing to me, why wouldn't it be allowed on numbers or strings?

Reply:
because it makes no sense to declare an argument as both optional and required unless it is nullable:

```rust
fn foo(#[required] param: Option<i32>) {
    ...
}
```

In theory this might be ok for `&str`, but currently we have a weird duality for strings - so I am not handling it just yet -- i.e. `#[required]` with strings can simply be expressed as `param: &str`.  In the future, once we only allow `Option<&str>`, this list will grow

Reply:
ok, so it would make sense for `&str` :-) I'm good, thanks for the explanation